### DISCUSSION TITLE
Unexpected performance of autocompletion

### DISCUSSION THREAD

**Discussion by gnuemacscoin** [minad/corfu#489] [Q&A] [answered]
There's this variable `corfu-auto-prefix` which I assumed would function like `company-minimum-prefix-length`, checking for minimum prefix length *before* querying the completion backend. When that backend is `lsp-mode`, in particular java's `jdtls`, this query can become slow, especially on short prefixes. In code I found that the check is performed only *after* querying the backend:
``` elisp
(defun corfu--capf-wrapper (fun &optional prefix)
  "Wrapper for `completion-at-point' FUN.
The wrapper determines if the Capf is applicable at the current
position and performs sanity checking on the returned result.
For non-exclusive Capfs wrapper additionally checks if the
current input can be completed successfully.  PREFIX is a prefix
length override, set to t for manual completion."
  (pcase (funcall fun)
    ((and res `(,beg ,end ,table . ,plist))
     (and (integer-or-marker-p beg) ;; Valid Capf result
          (<= beg (point) end)      ;; Sanity checking
          ;; When auto completing, check the prefix length!
          (let ((len (or prefix
                         (plist-get plist :company-prefix-length)
                         (- (point) beg))))
            (or (eq len t) (>= len corfu-auto-prefix)))
          ;; For non-exclusive Capfs, check for valid completion.
          (or (not (eq 'no (plist-get plist :exclusive)))
              (let* ((str (buffer-substring-no-properties beg end))
                     (pt (- (point) beg))
                     (pred (plist-get plist :predicate))
                     (md (completion-metadata (substring str 0 pt) table pred)))
                ;; We use `completion-try-completion' to check if there are
                ;; completions. The upstream `completion--capf-wrapper' uses
                ;; `try-completion' which is incorrect since it only checks for
                ;; prefix completions.
                (completion-try-completion str table pred pt md)))
          (cons fun res)))))
```
In documentation it is advised not to make `corfu-auto-delay` too short, acknowledging the performance impact of triggering autocompletions too frequently. I believe that this frequency could be more effectively reduced by just checking the prefix length ahead of time, for example like in
``` elisp
(defun corfu-auto-predicate ()
  (looking-back "[_[:alnum:]]\\{3,\\}")) ;; at least 3 letters, digits or underscores
(defun corfu--auto-post-command ()
  "Post command hook which initiates auto completion."
  (cancel-timer corfu--auto-timer)
  (if (and (not completion-in-region-mode)
           (not defining-kbd-macro)
           (not buffer-read-only)
           (corfu--match-symbol-p corfu-auto-commands this-command)
           (corfu--popup-support-p)
           (corfu-auto-predicate))
      (if (<= corfu-auto-delay 0)
          (corfu--auto-complete-deferred)
        (timer-set-time corfu--auto-timer
                        (timer-relative-time nil corfu-auto-delay))
        (timer-set-function corfu--auto-timer #'corfu--auto-complete-deferred
                            (list (corfu--auto-tick)))
        (timer-activate corfu--auto-timer))))
```

**Comment by minad** [comment]
> In documentation it is advised not to make corfu-auto-delay too short, acknowledging the performance impact of triggering autocompletions too frequently. I believe that this frequency could be more effectively reduced by just checking the prefix length ahead of time, for example like in

Yes, this would work in principle. However it is up to the Capf to determine the prefix. If a Capf is too slow initiating a completion session, it is a bug in the Capf - the initiation check must be fast by definition of the Capf protocol. Note that you could always wrap a slow Capf such that your cheap check is performed first, and only if the check succeeds continue with calling the slow Capf.

> In documentation it is advised not to make corfu-auto-delay too short, acknowledging the performance impact of triggering autocompletions too frequently. I believe that this frequency could be more effectively reduced by just checking the prefix length ahead of time, for example like in

The problem here is usually not the prefix check, but rather the actual computation of the candidates which will occur too often. As mentioned before, determining the prefix should not be costly.

**Reply by gnuemacscoin** [reply]
> Note that you could always wrap a slow Capf such that your cheap check is performed first, and only if the check succeeds continue with calling the slow Capf.

But sometimes I want to *manually* trigger an expensive completion, from what I understand wrapping would apply both to automatic and manual completions.

**Reply by minad** [reply, marked as answer]
Sure. For expensive manual completions my recommendation is to explicitly bind them to a separate key, see my Cape package and `cape-capf-interactive`. But my main point here is that the prefix computation alone should *never* be slow on the side of the Capf (in contrast to the candidate computation which may be more expensive). As you point out, we could use a potentially cheaper prefix check, but it should not matter, since the prefix check in the Capf should already be fast. You also imply that prefix handling for manual and automatic completion should be different, but this is better avoided since it will lead to inconsistencies in edge cases.
